homeassistant/components/onkyo/media_player.py

```diff
@@ -233,14 +237,35 @@ async def async_setup_platform(hass, config, async_add_entities, discovery_info=
 
     _LOGGER.debug("Provisioning Onkyo AVR device at %s:%d", host, port)
 
+    active_zones = []
+
+    def async_onkyo_discover_zones_callback(message):
+        """Receive the power status of the available zones on the AVR."""
+        zone, command, _ = message
+        if command in ["system-power", "power"]:
+            # When we receive the status for a zone, it is available on the AVR
+            # So we create an entity for the zone and continue discovery
+            if zone in zones:
+                _LOGGER.debug(
+                    "Discovered %s on %s, add to active zones", zones[zone], name
+                )
+                zone_entity = OnkyoAVR(
+                    avr, f"{name} {zones[zone]}", sources, zone, max_volume
+                )
+                active_zones.append(zone_entity)
+                async_add_entities([zone_entity])
+
     @callback
     def async_onkyo_update_callback(message):
         """Receive notification from transport that new data exists."""
         _LOGGER.debug("Received update callback from AVR: %s", message)
-        for zone in active_zones:
-            zone.process_update(message)
 
-    active_zones = []
+        _zone, _, _ = message
+        if any(active_zone.zone == _zone for active_zone in active_zones):
```

Comment:
I think `active_zones` could be a dict, and we simply look it up and call `process_update(message)` on the value, or `async_onkyo_discover_zones_callback(message)` if we can't find it. The could would be simpler, and we'll avoid sending messages to zones that don't need them